akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala

```diff
@@ -262,20 +292,19 @@ private[http] object StreamUtils {
    * INTERNAL API
    */
   @InternalApi
-  object CaptureMaterializationAndTerminationOp extends EntityStreamOp[(Future[Unit], Future[Unit], Future[Option[UniqueKillSwitch]])] {
-    def strictM: (Future[Unit], Future[Unit], Future[Option[UniqueKillSwitch]]) = (Future.successful(()), Future.successful(()), Future.successful(None))
-    def apply[T, Mat](source: Source[T, Mat]): (Source[T, Mat], (Future[Unit], Future[Unit], Future[Option[UniqueKillSwitch]])) = {
-      val materializationPromise = Promise[Unit]()
-      val killSwitchPromise = Promise[Option[UniqueKillSwitch]]()
-      val newSource0 =
-        source.via(Flow.fromGraph(KillSwitches.single[T]).mapMaterializedValue(ks => killSwitchPromise.trySuccess(Some(ks))))
-      val (newSource, completion) =
-        StreamUtils.captureTermination(newSource0.mapMaterializedValue { mat =>
-          materializationPromise.trySuccess(())
-          mat
-        })
-
-      (newSource, (materializationPromise.future, completion, killSwitchPromise.future))
+  object CaptureMaterializationAndTerminationOp extends EntityStreamOp[(Future[Any], Future[Unit], Option[KillSwitch])] {
+    def strictM: (Future[Any], Future[Unit], Option[KillSwitch]) = (Future.successful(()), Future.successful(()), None)
+    def apply[T, Mat](source: Source[T, Mat]): (Source[T, Mat], (Future[Any], Future[Unit], Option[KillSwitch])) = {
```

Comment:
Why not:
```suggestion
    def apply[T, Mat](source: Source[T, Mat]): (Source[T, Mat], (Future[Mat], Future[Unit], Option[KillSwitch])) = {
```
? Might make it clearer what it is even if we're not using the value ;)

Reply:
It won't work like this because it doesn't fit the `EntityStreamOp` supertype and also it doesn't actually return the materialized value.

But you are right that this result type is a bit opaque. I'll introduce a small case class for it for better readability.